File: Sources/TinySpectrum/WWBTimelineExporter.swift
import Foundation

enum WWBTimelineExporter {
    enum ExportError: LocalizedError {
        case notContinuous
        case noSamples
        case spanTooNarrow

        var errorDescription: String? {
            switch self {
            case .notContinuous: "Only continuous scans can be exported as a WWB timeline."
            case .noSamples: "The continuous scan does not contain any samples."
            case .spanTooNarrow: "Wireless Workbench requires a frequency span of at least 25 kHz."
            }
        }
    }

    private struct Grid {
        let startKHz: Int
        let stopKHz: Int
        let stepKHz: Int

        var count: Int { (stopKHz - startKHz) / stepKHz + 1 }
        var recordSize: Int { 4 + 4 + 4 + count * 2 + 2 }
    }

    static func data(for scan: SpectrumScan, title: String) throws -> Data {
        guard let captures = scan.captures else { throw ExportError.notContinuous }
        guard let first = captures.first, first.points.count >= 2 else { throw ExportError.noSamples }
        let grid = try makeGrid(from: first.points)
        let safeTitle = title.isEmpty ? "TinySpectrum continuous scan" : title

        func curve(name: String, color: String) -> [String: Any] {
            [
                "Color": color,
                "CoordinationSource": true,
                "FreqRanges": [[
                    "EndFreq": grid.stopKHz,
                    "StartFreq": grid.startKHz,
                    "StepFreq": grid.stepKHz
                ]],
                "Name": name,
                "ResolutionBandWidth": resolutionBandwidthKHz(scan.rbw)
            ]
        }
        let header: [String: Any] = [
            "AmplUnits": "dBm",
            "BinarySchema": [
                ["Bytes": 4, "DataValue": "start-of-sweep"],
                ["Bytes": 4, "DataValue": "id"],
                ["Bytes": 4, "DataValue": "timestamp"],
                ["Curve": curve(name: "Antenna A", color: "#ffff00")],
                ["Bytes": 2, "DataValue": "crc16"]
            ],
            "BitWidth": 16,
            "FreqUnits": "KHz",
            "NoDataValue": -1400,
            "PeriodicSpecialSweeps": [
                ["Interval": 40, "Type": "Periodic running peakhold"],
                ["Interval": 100, "Type": "Periodic interval peakhold"]
            ],
            "Scale Factor": 10,
            "ScannerModel": "",
            "ScannerName": "AD600",
            "StartDate": dateString(first.date),
            "StartTime": timeString(first.date),
            "Title": safeTitle,
            "Version": "1.0.0.0"
        ]

        var output = Data("//@ShureScan\n".utf8)
        output.append(try JSONSerialization.data(withJSONObject: header, options: [.prettyPrinted, .sortedKeys]))
        output.append(Data("\n@Binary:".utf8))

        var runningPeak = [Int16](repeating: -1400, count: grid.count)
        var intervalPeak = [Int16](repeating: -1400, count: grid.count)
        var totalRecords: UInt32 = 0
        let startDate = first.date

        for (offset, capture) in captures.enumerated() {
            let samples = resample(capture.points, onto: grid)
            mergePeak(samples, into: &runningPeak)
            mergePeak(samples, into: &intervalPeak)
            let identifier = UInt32(offset + 1)
            let elapsed = max(0, capture.date.timeIntervalSince(startDate).rounded())
            output.append(record(id: identifier, timestamp: UInt32(clamping: Int(elapsed)), samples: samples))
            totalRecords += 1

            if identifier.isMultiple(of: 40) {
                output.append(record(id: 0, timestamp: 0, samples: runningPeak))
                totalRecords += 1
            }
            if identifier.isMultiple(of: 100) {
                output.append(record(id: 0, timestamp: 0, samples: intervalPeak))
                totalRecords += 1
                intervalPeak = [Int16](repeating: -1400, count: grid.count)
            }
        }

        let extended: [String: Any] = [
            "Band": "Wideband",
            "Creator": "TinySpectrum 2.0.1-beta.2",
            "ScanName": [safeTitle],
            "UserCurveColors": ["#ffff00"]
        ]
        output.append(Data("@Extended:\n".utf8))
        output.append(try JSONSerialization.data(withJSONObject: extended, options: [.prettyPrinted, .sortedKeys]))
        output.append(0x0A)
        output.append(Data(repeating: 0x20, count: max(0, grid.recordSize - 6)))
        output.append(0x0A)
        output.append(Data("@End".utf8))
        output.appendBigEndian(UInt32(captures.count))
        output.appendBigEndian(totalRecords)
        return output
    }

    private static func makeGrid(from points: [ScanPoint]) throws -> Grid {
        let sorted = points.sorted { $0.frequency < $1.frequency }
        guard let first = sorted.first, let last = sorted.last else { throw ExportError.noSamples }
        let startKHz = Int(ceil(first.frequency / 1_000))
        let availableStopKHz = Int(floor(last.frequency / 1_000))
        guard availableStopKHz - startKHz >= 25 else { throw ExportError.spanTooNarrow }
        let sourceStepKHz = (last.frequency - first.frequency) / Double(max(1, sorted.count - 1)) / 1_000
        let stepKHz = max(25, Int(sourceStepKHz.rounded()))
        let count = max(2, (availableStopKHz - startKHz) / stepKHz + 1)
        return Grid(startKHz: startKHz, stopKHz: startKHz + (count - 1) * stepKHz, stepKHz: stepKHz)
    }

    private static func resample(_ points: [ScanPoint], onto grid: Grid) -> [Int16] {
        let sorted = points.sorted { $0.frequency < $1.frequency }
        guard sorted.count >= 2 else { return [Int16](repeating: -1400, count: grid.count) }
        var result: [Int16] = []
        result.reserveCapacity(grid.count)
        var upperIndex = 1
        for index in 0..<grid.count {
            let frequency = Double(grid.startKHz + index * grid.stepKHz) * 1_000
            while upperIndex < sorted.count - 1, sorted[upperIndex].frequency < frequency { upperIndex += 1 }
            let lower = sorted[upperIndex - 1]
            let upper = sorted[upperIndex]
            guard frequency >= lower.frequency, frequency <= upper.frequency else {
                result.append(-1400)
                continue
            }
            let span = max(1, upper.frequency - lower.frequency)
            let fraction = (frequency - lower.frequency) / span
            let level = lower.level + (upper.level - lower.level) * fraction
            result.append(Int16(clamping: Int((level * 10).rounded())))
        }
        return result
    }

    private static func mergePeak(_ samples: [Int16], into peak: inout [Int16]) {
        for index in samples.indices where samples[index] != -1400 {
            peak[index] = max(peak[index], samples[index])
        }
    }

    private static func record(id: UInt32, timestamp: UInt32, samples: [Int16]) -> Data {
        var value = Data("@Swp".utf8)
        value.appendBigEndian(id)
        value.appendBigEndian(timestamp)
        for sample in samples { value.appendBigEndian(sample) }
        value.appendBigEndian(crc16ARC(value))
        return value
    }

    private static func crc16ARC(_ data: Data) -> UInt16 {
        var crc: UInt16 = 0
        for byte in data {
            crc ^= UInt16(byte)
            for _ in 0..<8 { crc = (crc & 1) == 1 ? (crc >> 1) ^ 0xA001 : crc >> 1 }
        }
        return crc
    }

    private static func resolutionBandwidthKHz(_ label: String) -> Double {
        let number = Scanner(string: label).scanDouble() ?? 30
        return label.lowercased().contains("khz") ? number : number / 1_000
    }

    private static func dateString(_ date: Date) -> String {
        let formatter = DateFormatter()
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.dateFormat = "MM/dd/yyyy"
        return formatter.string(from: date)
    }

    private static func timeString(_ date: Date) -> String {
        let formatter = DateFormatter()
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.dateFormat = "HH:mm:ss"
        return formatter.string(from: date)
    }
}

private extension Data {
    mutating func appendBigEndian<T: FixedWidthInteger>(_ value: T) {
        var bigEndian = value.bigEndian
        Swift.withUnsafeBytes(of: &bigEndian) { append(contentsOf: $0) }
    }
}

File: Windows/TinySpectrum.Windows/WwbTimelineExporter.cs
using System.Buffers.Binary;
using System.Globalization;
using System.Text;
using System.Text.Json;

namespace TinySpectrum.Windows;

public static class WwbTimelineExporter
{
    private static readonly JsonSerializerOptions JsonOptions = new() { WriteIndented = true };

    public static async Task WriteCsvAsync(IReadOnlyList<ScanPoint> points, Stream output)
    {
        await using var writer = new StreamWriter(output, new UTF8Encoding(false), leaveOpen: true);
        foreach (var point in points)
            await writer.WriteLineAsync(string.Create(CultureInfo.InvariantCulture, $"{point.Frequency / 1e6:F6},{point.Level:F2}"));
        await writer.FlushAsync();
    }

    public static async Task WriteAsync(SpectrumScan scan, Stream output)
    {
        if (scan.Captures is not { Count: > 0 } captures) throw new InvalidOperationException("Only continuous scans have a WWB timeline.");
        var grid = MakeGrid(captures[0].Points);
        var title = $"TinySpectrum {scan.Date:yyyy-MM-dd HH-mm}";
        var curve = new Dictionary<string, object>
        {
            ["Color"] = "#ffff00",
            ["CoordinationSource"] = true,
            ["FreqRanges"] = new[] { new Dictionary<string, object> { ["EndFreq"] = grid.StopKHz, ["StartFreq"] = grid.StartKHz, ["StepFreq"] = grid.StepKHz } },
            ["Name"] = "Antenna A",
            ["ResolutionBandWidth"] = ParseRbwKHz(scan.Rbw)
        };
        var header = new Dictionary<string, object>
        {
            ["AmplUnits"] = "dBm",
            ["BinarySchema"] = new object[]
            {
                new Dictionary<string, object> { ["Bytes"] = 4, ["DataValue"] = "start-of-sweep" },
                new Dictionary<string, object> { ["Bytes"] = 4, ["DataValue"] = "id" },
                new Dictionary<string, object> { ["Bytes"] = 4, ["DataValue"] = "timestamp" },
                new Dictionary<string, object> { ["Curve"] = curve },
                new Dictionary<string, object> { ["Bytes"] = 2, ["DataValue"] = "crc16" }
            },
            ["BitWidth"] = 16,
            ["FreqUnits"] = "KHz",
            ["NoDataValue"] = -1400,
            ["PeriodicSpecialSweeps"] = new object[]
            {
                new Dictionary<string, object> { ["Interval"] = 40, ["Type"] = "Periodic running peakhold" },
                new Dictionary<string, object> { ["Interval"] = 100, ["Type"] = "Periodic interval peakhold" }
            },
            ["Scale Factor"] = 10,
            ["ScannerModel"] = "",
            ["ScannerName"] = "AD600",
            ["StartDate"] = captures[0].Date.ToString("MM/dd/yyyy", CultureInfo.InvariantCulture),
            ["StartTime"] = captures[0].Date.ToString("HH:mm:ss", CultureInfo.InvariantCulture),
            ["Title"] = title,
            ["Version"] = "1.0.0.0"
        };
        await output.WriteAsync(Encoding.UTF8.GetBytes("//@ShureScan\n"));
        await JsonSerializer.SerializeAsync(output, header, JsonOptions);
        await output.WriteAsync(Encoding.UTF8.GetBytes("\n@Binary:"));

        var runningPeak = Enumerable.Repeat((short)-1400, grid.Count).ToArray();
        var intervalPeak = Enumerable.Repeat((short)-1400, grid.Count).ToArray();
        uint totalRecords = 0;
        for (var captureIndex = 0; captureIndex < captures.Count; captureIndex++)
        {
            var samples = Resample(captures[captureIndex].Points, grid);
            MergePeak(samples, runningPeak); MergePeak(samples, intervalPeak);
            var id = (uint)captureIndex + 1;
            var timestamp = (uint)Math.Max(0, Math.Round((captures[captureIndex].Date - captures[0].Date).TotalSeconds));
            await output.WriteAsync(Record(id, timestamp, samples)); totalRecords++;
            if (id % 40 == 0) { await output.WriteAsync(Record(0, 0, runningPeak)); totalRecords++; }
            if (id % 100 == 0)
            {
                await output.WriteAsync(Record(0, 0, intervalPeak)); totalRecords++;
                Array.Fill(intervalPeak, (short)-1400);
            }
        }

        var extended = new Dictionary<string, object>
        {
            ["Band"] = "Wideband",
            ["Creator"] = "TinySpectrum Windows 0.1.0-beta.1",
            ["ScanName"] = new[] { title },
            ["UserCurveColors"] = new[] { "#ffff00" }
        };
        await output.WriteAsync(Encoding.UTF8.GetBytes("@Extended:\n"));
        await JsonSerializer.SerializeAsync(output, extended, JsonOptions);
        await output.WriteAsync(new byte[] { 0x0A });
        await output.WriteAsync(Enumerable.Repeat((byte)0x20, Math.Max(0, grid.RecordSize - 6)).ToArray());
        await output.WriteAsync(Encoding.UTF8.GetBytes("\n@End"));
        var counters = new byte[8];
        BinaryPrimitives.WriteUInt32BigEndian(counters.AsSpan(0, 4), (uint)captures.Count);
        BinaryPrimitives.WriteUInt32BigEndian(counters.AsSpan(4, 4), totalRecords);
        await output.WriteAsync(counters);
    }

    private static Grid MakeGrid(IReadOnlyList<ScanPoint> points)
    {
        if (points.Count < 2) throw new InvalidOperationException("A WWB timeline needs at least two points.");
        var sorted = points.OrderBy(x => x.Frequency).ToArray();
        var start = (int)Math.Ceiling(sorted[0].Frequency / 1000);
        var availableStop = (int)Math.Floor(sorted[^1].Frequency / 1000);
        if (availableStop - start < 25) throw new InvalidOperationException("Wireless Workbench requires at least a 25 kHz span.");
        var sourceStep = (sorted[^1].Frequency - sorted[0].Frequency) / (sorted.Length - 1) / 1000;
        var step = Math.Max(25, (int)Math.Round(sourceStep));
        var count = Math.Max(2, (availableStop - start) / step + 1);
        return new(start, start + (count - 1) * step, step);
    }

    private static short[] Resample(IReadOnlyList<ScanPoint> source, Grid grid)
    {
        var sorted = source.OrderBy(x => x.Frequency).ToArray();
        var result = new short[grid.Count];
        var sourceIndex = 0;
        for (var i = 0; i < result.Length; i++)
        {
            var frequency = (grid.StartKHz + i * grid.StepKHz) * 1000.0;
            while (sourceIndex + 1 < sorted.Length && sorted[sourceIndex + 1].Frequency < frequency) sourceIndex++;
            double level;
            if (frequency <= sorted[0].Frequency) level = sorted[0].Level;
            else if (frequency >= sorted[^1].Frequency) level = sorted[^1].Level;
            else
            {
                var a = sorted[sourceIndex]; var b = sorted[Math.Min(sourceIndex + 1, sorted.Length - 1)];
                var ratio = (frequency - a.Frequency) / Math.Max(1, b.Frequency - a.Frequency);
                level = a.Level + (b.Level - a.Level) * ratio;
            }
            result[i] = (short)Math.Clamp((int)Math.Round(level * 10), short.MinValue, short.MaxValue);
        }
        return result;
    }

    private static byte[] Record(uint id, uint timestamp, IReadOnlyList<short> samples)
    {
        var data = new byte[12 + samples.Count * 2 + 2];
        BinaryPrimitives.WriteUInt32BigEndian(data.AsSpan(0, 4), 0xFFFFFFFF);
        BinaryPrimitives.WriteUInt32BigEndian(data.AsSpan(4, 4), id);
        BinaryPrimitives.WriteUInt32BigEndian(data.AsSpan(8, 4), timestamp);
        for (var i = 0; i < samples.Count; i++) BinaryPrimitives.WriteInt16BigEndian(data.AsSpan(12 + i * 2, 2), samples[i]);
        BinaryPrimitives.WriteUInt16BigEndian(data.AsSpan(data.Length - 2), Crc16(data.AsSpan(0, data.Length - 2)));
        return data;
    }

    public static ushort Crc16(ReadOnlySpan<byte> data)
    {
        ushort crc = 0;
        foreach (var value in data)
        {
            crc ^= value;
            for (var bit = 0; bit < 8; bit++) crc = (ushort)((crc & 1) == 1 ? (crc >> 1) ^ 0xA001 : crc >> 1);
        }
        return crc;
    }

    private static void MergePeak(IReadOnlyList<short> samples, short[] peak)
    {
        for (var i = 0; i < Math.Min(samples.Count, peak.Length); i++) if (samples[i] > peak[i]) peak[i] = samples[i];
    }

    private static double ParseRbwKHz(string label)
    {
        var token = label.Split(' ', StringSplitOptions.RemoveEmptyEntries)[0];
        var value = double.TryParse(token, NumberStyles.Float, CultureInfo.InvariantCulture, out var parsed) ? parsed : 30;
        return label.Contains("kHz", StringComparison.OrdinalIgnoreCase) ? value : value / 1000;
    }

    private sealed record Grid(int StartKHz, int StopKHz, int StepKHz)
    {
        public int Count => (StopKHz - StartKHz) / StepKHz + 1;
        public int RecordSize => 12 + Count * 2 + 2;
    }
}
